additional brick layout options

frazazel opened the pull request:
Adds additional formatting options for layouts, allowing users to have rows and columns of bricks, and set widths for bricks or groups of bricks.
chit.__.layout can now use (){}|, for defining row/column configuration of bricks, and :_%, :_* or :_px for defining the width of bricks or groups of bricks.

examples:
set chit.roof.layout=character|gear:112px,stats:3*|{familiar,trail}:2*
() define horizontally aligned groups of bricks
{} define vertically alligned groups of bricks
| defines bricks horizontally adjacent to other bricks
, defines a new row of bricks
:x% means that the preceding brick or group of bricks will have a width equal to x%.
:x* means that the preceding brick or group of bricks will have a width weighted by x, compared to other bricks in the same row (defaults to weight of 1 if not set).
:xpx means that the preceding brick or group of bricks will have a width equal to x pixels.
Existing chit.__.layout strings work as usual without needing to change.

Page html includes additional div containers for each brick and each group of bricks, and width is set on these div containers. Minor changes to CSS margins, etc. are expected from this change.

# Description

Please include a summary of the change.

Fixes # (issue)

## Screenshots

Please include a screenshot both with and without your change active.

## Checklist:

- [X] I have performed a self-review of my own code.
- [X] I have commented my code, particularly in hard-to-understand areas.
- [X] I have based by pull request against the [main branch](https://github.com/Loathing-Associates-Scripting-Society/ChIT/tree/main) or have a good reason not to.
<img width="203" alt="ChIT revisions" src="https://user-images.githubusercontent.com/13992/162604709-d3d9c6e2-0a4a-4c5e-a06b-ad70ecf54389.PNG">

soolar commented on src/relay/charpane.ash:
```diff
@@ -4412,26 +4410,234 @@ void bakeBricks() {
+	buffer buff;
+	string brick;
+	int totalStars = starCount(styleInfos) + count(bricks) - count(styleInfos);
+	//append(buff, '<div class="brick_row">');
```
If this is no longer necessary, please remove it (or is it just something you occasionally uncommented to help test with something for some reason? If that's the case, please add another comment explaining the use case there).

frazazel replied:
It's no longer necessary. Removing it.

soolar commented on src/relay/charpane.ash:
```diff
@@ -4412,26 +4410,234 @@ void bakeBricks() {
+			append(buff, '<div class="brick_holder" '+stylize(styleInfos[i],totalStars)+'>'+brick+'</div>');
+		}
+	}
+	//append(buff, '</div>');
```
Same as above commented line

frazazel replied:
Got it.

soolar commented on src/relay/charpane.ash:
```diff
@@ -4412,26 +4410,234 @@ void bakeBricks() {
+	int[int] j = {0:0};
+	foreach n,s in tokens {
+		if(read_style){
+			styleInfo[pLevel,i[pLevel]] = tokens[n];
```
Wouldn't `s` work just fine instead of `tokens[n]`?

frazazel replied:
Should be. This is a probably artefact of a previous version of the algorithm.

soolar reviewed (changes requested):
This looks absolutely awesome. The only major issue (not one of my comments below, those are all very minor) is the lack of README update to explain the extra layout options. I'd rather have you write that bit than me based on my (possibly incorrect in some ways) interpretation of your code and description.

Also, one thing that would be very nice is some error handling, but I can probably add that myself if you don't want to bother. Just basic stuff like "you forgot a closing parenthesis you dork" as an error message instead of what I imagine will happen (a really messed up layout with no super clear reason).

Anyway, I'm going to spend a little bit of time playing around with layouts and see if I find any other issues.

soolar commented:
Ah, found an issue: Bricks that sometimes don't display (terminal, songboom, horsery, robo, probably some others) cause issues if you try to squeeze them in a row with something else. I think better behavior there would be to behave as if that portion of the layout didn't exist. (ie if I have `thrall|horsery,robo|terminal` for some reason but am not in You, Robot atm, it should behave as if that area was just `thrall|horsery,terminal`. Fortunately the bricks are all baked in advance, so you can just check `chitBricks[whatever] == ""`.

frazazel commented:
New commit pushed.

I've added basic error handling, where mismatched parens display an error message within the ChIT panel, and also closes all still-open parens. This is ugly, but users will at least be motivated to fix their errors.
I've refactored the way that addBricks looks up bricks within chitBricks, and I have it now ignoring empty bricks and their specified widths.
I've also completed the other changes you asked for, and I've updated the README.
Also, after further use I have noticed that I can't find a real use for explicit horizontal groups of bricks, which is already implied by the basic formatting. As such, I'm opting to change the ( ) and { } around, so that the more common vertical groups of bricks get the more common ( ) parens, and explicit horizontal groups of bricks are still available via { } curly braces for the 2 people who someday discover they need this feature. You may need to update your chit.__.layout preferences.

soolar commented on src/relay/charpane.ash:
```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
+					case "footer":
+						break;	//Special Bricks that are inserted manually in the correct places
+					default:
+						if(chitBricks contains s && chitBricks[s] != "") {
```
Not really important, but I believe the `chitBricks contains s` check isn't needed, because maps in ASH return the default value for their output type when indexed with an unused value.

frazazel replied:
Okay, I'll test to make sure that works, and delete the unneeded code.

soolar commented on src/relay/charpane.ash:
```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
+				}
 			}
 		}
+		if(pLevel != 0)
```
Should this be `pLevel == 0`? I'm not 100% sure about that, but considering the next thing done is a loop UNTIL it is 0, that seems off to me?

frazazel replied:
No, this is correct. pLevel counts up as open parens are found, and counts down when closing parens are found. If it reaches 0 at the end, then there are the same number of opening and closing parens. It writes an error if pLevel is not 0. 
If too insufficient parens are closed, then some bricks are not written to the charpane, so simple error correction is to close all still-open parens, so that the bricks requested are at least showing up, but possibly not in the configuration expected, because the user made an error.

soolar commented on src/relay/charpane.ash:
```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
 //receive chit.__.layout string, tokenize it, and send it for processing.
 	string[int] tokens = tokenize(layout, $strings[\,,|,(,),:,{,}]);
-	return addBricks(tokens);
+	return addBricksHelper(tokens);
```
Considering this helper is only used in one place as far as I can see, it doesn't really seem needed... But again, that's not a big deal at all.

frazazel replied:
It was more work to refactor it. If I wrote it again, I would probably not write it this way. I pulled in into the main addBricks function so that it has access to the "layout" parameter (used in error message) without having to pass it along as another parameter. No strong feelings about whether it needs to change, so I'm inclined to leave it.

soolar replied:
Fair enough!

soolar commented on src/relay/charpane.ash:
```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
 			}
 		}
+		if(pLevel != 0)
+			append(result, "<span>Malformed chit layout (mismatched parentheses): " + layout + "<span>");
```
I was thinking more along the line of calling `error` than embedding the error straight in to the resulting charpane, but... I kind of like this more, so that's cool.

Second `<span>` should be `</span>` though.

frazazel replied:
Yeah, I wrote this so I could see it working before I figured out how to throw an error message, and then I kinda liked it. It shows people which room the error is in, and if you're in a pre-rollover panic, you can ignore it for now and fix it later. I think it should probably get a <div class="error_message"> to wrap it, and then add some basic styling to the css file. I'll add that, and fix the </span> error.

soolar replied:
Looks nicer, but seems to no longer actually display the string value for some reason.

frazazel replied:
Huh. I styled the string using "font-family: monospace" to give the a code-y look. Did it overflow to the next line, and then being cut off? I had it working on my machine.

soolar commented on src/relay/charpane.ash:
```diff
@@ -4532,112 +4532,112 @@ buffer addGroup(string[int] rowHTMLs, string className) {
-		if(length(brick) > 0) {
-			append(buff, '<div class="brick_holder" '+stylize(styleInfos[i],totalStars)+'>'+brick+'</div>');
-		}
+		append(buff, '<div class="brick_holder" '+stylize(styleInfos[i],totalStars)+'>'+s+'</div>');
```
Minor thing I didn't notice before: in most cases ChIT will do `buff.append(whatever)` instead of `append(buff, whatever)`. I'm not particularly bothered by things being done differently in some places though.

frazazel replied:
I'm not sure if you're actually asking for this to be changed, but since you made the comment I'll take a look. Consistent style choices are nice.

frazazel replied:
I actually tried to change it from buff.append(whatever), and mafia started giving me errors. Not sure what's going on, but it doesn't look like this change works.

soolar replied:
Weird. Well, it's not a big issue, so that's fine.

soolar reviewed (changes requested):
Only thing here that's both important and I'm sure about is the double opening `<span>` thing.

soolar commented:
What exactly is the difference between horizontal groups and just using `|` between two bricks anyway? I'm not 100% clear on that. Is it necessary if you want a horizontal group within a vertical group, or something like that?

soolar commented:
Also man, I've been working on a rewrite of ChIT in React, and it's funny how much easier this work would be there. It's not QUITE as simple as this example because I think some styling would be necessary to prevent a ton of blank space in the case of bricks of uneven size, but:

```javascript
<VStack>
  <HStack>
    <CharacterBrick />
    <GearBrick maxW="112px" />
  </HStack>
  <HStack>
    <StatsBrick />
    <VStack>
      <FamiliarBrick />
      <TrailBrick />
    </VStack>
  </HStack>
  <EffectsBrick />
</VStack>
```
Would mostly reproduce your example screenshot. Well, only once all of those bricks are actually finished being written in the rewrite, of course, haha.

soolar commented:
Ah, found one more issue: If a `()` or `{}` group has nothing but empty bricks in it, it'll still add the wrapper, resulting in empty space. This may be a somewhat niche scenario but I actually found it while setting up my charpane to be even more complicated (with certain bricks I'll only have in non-standard scenarios).

soolar commented:
Hmm, something else, although I THINK it's harmless? I set up my florist brick next to the trail brick like this `{trail,florist:100px}` (within a longer string) and when florist is empty (because of not having access during community service) the `100px` makes it in to the `default` branch of the `switch` within `addBricksHelper` (I added `print(s);` at the start of the `default` case before `if(chitBricks contains s && chitBricks[s] != "") {` to see if my brick names were even showing up in the right place and that revealed this). Like, pretty definitely harmless because it'll just fail that check, but still, just thought you should know.

Actually, something at least tangentially related at the very least is harmful. I'm experimenting with a pretty wild `chit.roof.layout` of `{character,stats,trail|florist:100px,helpers}|gear:45px` (to make the gear brick totally vertical with one equip slot per line, similar to the normal equipment page). Unfortunately, when the florist brick is empty (due to no access to florists such as in community service), this causes everything but the gear brick to disappear.

Here's what that looks like when it is working though, in case you're curious (I just hardcoded the florist brick to always display for convenience's sake).
![image](https://user-images.githubusercontent.com/1923854/163054785-324d918e-368b-4705-a5d6-ca8e63caeb9a.png)
Also, one thing I'm not a massive fan of is that the padding between the gear brick and everything else in its block, and between that whole block and the walls is something like twice as long as between, say, the character and stats bricks.

frazazel commented:
New commit pushed.

I wasn't able to reproduce your error with the Florist Friars breaking the whole display, but I tested that late after addressing the other issues. Let me know if it's still an issue, and I'll look into it harder.

I've slightly reworked how styles are being assigned to blocks, so that fewer <div>s are needed. I also reworked the CSS to have more consistent spacing. There should be no gradual buildup of margins as more and more groups are placed around bricks.

soolar commented:
I still get the florist issue. `chit.roof.layout` is `{character,stats,trail|florist:100px}|gear:45px,helpers` atm, and this is what it looks like if florist isn't available (or I just change the `florist_available()` check in `bakeFlorist` to `false` while it is available.

![image](https://user-images.githubusercontent.com/1923854/163892183-a4e103b2-49d3-493d-bad8-0b8f44951450.png)

It could be because `florist` is touching a `|`, because the same type of issue didn't happen with `helpers` when `chit.roof.layout` was `{character,stats,trail|florist:100px,helpers}|gear:45px` and `helpers` was blank.

The padding fix definitely makes it look a lot nicer when it is working though!

If you still can't repro the empty brick issue I'm having with florist, let me know. I'll probably just merge it in that case and go ahead and investigate it myself.

frazazel commented:
Okay, I was testing with a different brick that wasn't available. I'll try testing with that specific string, and see if I can figure out what's going wrong.

frazazel commented:
Alright, new commit pushed.

I found the issue, and the example you provided works as expected now.
